File: java/blackjack/blackjack.java
package blackjack;

import java.util.*;
import java.util.concurrent.ThreadLocalRandom;

import javax.swing.text.DefaultStyledDocument.ElementSpec;

import blackjack.util.util;
public class blackjack{
    private ArrayList<Integer> bankerHand;
    private ArrayList<Integer> playerHand;
    private boolean debug;
    private ArrayList<Integer> bankerHandSum;
    private ArrayList<Integer> playerHandSum;
    private int playerChip;
    private int bet;
    private boolean stand;
    private boolean ended;
    private ArrayList<Integer> deck;
    
    public blackjack(boolean debug){
        this.bankerHand=new ArrayList<Integer>();
        this.playerHand=new ArrayList<Integer>();
        this.debug=debug;
        this.bankerHandSum=new ArrayList<Integer>();
        this.playerHandSum=new ArrayList<Integer>();
        this.playerChip=100;
        this.bet=20;
        this.stand=false;
        this.ended=false;
        this.deck=new ArrayList<Integer>();
        for(int i=1;i<=13;i++){
            for(int j=0;j<4;j++){
                this.deck.add(i);
            }
        }
        for(int i=0;i<10;i++){
            Collections.shuffle(this.deck);
        }
    }
    
    public void continueGame(boolean newDeck){
        this.bankerHand=new ArrayList<Integer>();
        this.playerHand=new ArrayList<Integer>();
        this.bankerHandSum=new ArrayList<Integer>();
        this.playerHandSum=new ArrayList<Integer>();
        this.bet=20;
        this.stand=false;
        this.ended=false;
        if(newDeck){
            this.deck=new ArrayList<Integer>();
            for(int i=1;i<=13;i++){
                for(int j=0;j<4;j++){
                    this.deck.add(i);
                }
            }
            for(int i=0;i<10;i++){
                Collections.shuffle(this.deck);
            }
        }
    }

    public String print(){
        StringBuilder m=new StringBuilder("=========================\n");
        m.append("    Banker's Hand:");
        if(this.stand==false){
            m.append(" ?");
        }else{
            m.append(" "+util.cardConverter(this.bankerHand.get(0)));
        }
        for(int i=1;i<this.bankerHand.size();i++){
            m.append(" "+util.cardConverter(this.bankerHand.get(i)));
        }
        m.append("\n  > Banker's Sum:");
        for(int i=0;i<this.bankerHandSum.size();i++){
            m.append(" "+this.bankerHandSum.get(i));
        }
        m.append("\n    Player's Hand:");
        for(int i=0;i<this.playerHand.size();i++){
            m.append(" "+this.playerHand.get(i));
        }
        m.append("\n  > Player's Sum:");
        for(int i=0;i<this.playerHandSum.size();i++){
            m.append(" "+this.playerHandSum.get(i));
        }
        m.append("\n    Player's Remaining chip: "+this.playerChip);
        m.append("\n    Bet: "+this.bet);
        if(this.debug){
            m.append("\n    Deck length: "+this.deck.size());
            m.append("\n    Deck:");
            for(int i=0;i<this.deck.size();i++){
                m.append(" "+util.cardConverter(this.deck.get(i)));
            }
        }
        m.append("\n=========================\n");
        System.out.println(m);
        return m.toString();
    }

    public ArrayList<Integer> getPlayerHand(){
        return this.playerHand;
    }
    
    public ArrayList<Integer> getBankerHand(){
        if(this.stand){
            return this.bankerHand;
        }else{
            ArrayList<Integer> r = new ArrayList<Integer>();
            r.add(-1);
            r.add(this.bankerHand.get(1));
            return r;
        }
    }

    public boolean getEnded(){
        return this.ended;
    }

    private void calcBankerSum(){
        if(this.stand){
            ArrayList<Integer> sums=new ArrayList<Integer>();
            sums.add(0);
            for(int i=0;i<this.bankerHand.size();i++){
                int card=this.bankerHand.get(i);
                if(card==11 || card==12 || card==13){
                    for(int j=0;j<sums.size();j++){
                        sums.set(j,sums.get(j)+10);
                    }
                }else if(card!=1){
                    for(int j=0;j<sums.size();j++){
                        sums.set(j,sums.get(j)+card);
                    }
                }else{
                    int old_size=sums.size();
                    for(int j=0;j<old_size;j++){
                        sums.add(sums.get(j)+1);
                        sums.add(sums.get(j)+11);
                    }
                }
            }
            this.bankerHandSum.clear();
            Collections.sort(sums);
            this.bankerHandSum.add(0);
            if(sums.size()>1){
                if (sums.get(1)<=21){
                    this.bankerHandSum.add(1);
                }
            }
        }
    }

    private void calcPlayerSum(){
        ArrayList<Integer> sums=new ArrayList<Integer>();
        sums.add(0);
        for(int i=0;i<this.playerHand.size();i++){
            int card=this.playerHand.get(i);
            if(card==11 || card==12 || card==13){
                for(int j=0;j<sums.size();j++){
                    sums.set(j,sums.get(j)+10);
                }
            }else if(card!=1){
                for(int j=0;j<sums.size();j++){
                    sums.set(j,sums.get(j)+card);
                }
            }else{
                int old_size=sums.size();
                for(int j=0;j<old_size;j++){
                    sums.add(sums.get(j)+1);
                    sums.add(sums.get(j)+11);
                }
            }
        }
        this.playerHandSum.clear();
        Collections.sort(sums);
        this.playerHandSum.add(0);
        if(sums.size()>1){
            if (sums.get(1)<=21){
                this.playerHandSum.add(1);
            }
        }
    }
    
    private void bankerHit(){
        this.bankerHand.add(this.deck.remove(0));
        this.calcBankerSum();
        if(this.bankerHandSum.get(0)!=-1 && Collections.max(this.bankerHandSum)>=21){
            this.judge();
        }
    }

    public void playerHit(){
        this.playerHand.add(this.deck.remove(0));
        this.calcPlayerSum();
        if(this.playerHand
    }
    
    public void playerDouble(){

    }

    private void judge(){

    }   
}

File: python/blackjack.py
import random

class blackjack():
    def __init__(self,debug=False):
        print("\n\n***** New Player with new shuffled deck *****")
        self.bankerHand=[]
        self.playerHand=[]
        self.debug=debug
        self.bankerHandSum='?'
        self.playerHandSum='?'
        self.playerChip=100
        self.bet=20
        self.stand=False
        self.ended=False
        self.gameResult=0
        self.deck=[i+1 for i in range(13) for j in range(4)]
        for i in range(5):
            self.shuffleDeck()
        for i in range(2):
            self.bankerHand.append(self.hit())
            self.playerHand.append(self.hit())
        self.calcPlayerSum()
        print(self.getMsg())
        if max(self.playerHandSum)==21:
            self.judge()
            
    def newPlayer(self):
        self.__init__()

    def continueGame(self,newDeck=False):
        self.bankerHand=[]
        self.playerHand=[]
        self.bankerHandSum='?'
        self.playerHandSum='?'
        self.bet=20
        self.stand=False
        self.ended=False
        self.gameResult=0
        if(self.playerChip<self.bet):
            print("Chip not enough for bet, end game.")
            self.ended=True
            return
        if newDeck:
            print("\n\n*** Continue with new deck ***")
            self.deck=[i+1 for i in range(13) for j in range(4)]
            for i in range(5):
                self.shuffleDeck()
        else:
            print("\n\n*** Continue with original deck ***")
        for i in range(2):
            self.bankerHand.append(self.hit())
            self.playerHand.append(self.hit())
        self.calcPlayerSum()
        print(self.getMsg())
        if max(self.playerHandSum)==21:
            self.judge()
       

    def shuffleDeck(self):
        if self.debug:
            print("Shuffling deck")
        random.shuffle(self.deck)

    def getDeck(self):
        return self.deck

    def getBankerHand(self):
        return self.bankerHand

    def getPlayerHand(self):
        return self.playerHand

    def getPlayerChip(self):
        return self.playerChip

    def getBankerHandSum(self):
        self.calcBankerSum()
        return self.bankerHandSum
    
    def getPlayerHandSum(self):
        self.calcPlayerSum()
        return self.playerHandSum

    def getStatus(self):
        m="B["+("?" if self.stand==False else str(self.bankerHand[0]))
        for card in self.bankerHand[1:]:
            m+=(","+str(card) if (card!=1 and card<11) else(",A" if card==1 else(",J" if card==11 else(",Q" if card==12 else ",K"))))
        m+="]"
        m+=("P["+(str(self.playerHand[0]) if (self.playerHand[0]!=1 and self.playerHand[0]<11) else("A" if self.playerHand[0]==1 else("J" if self.playerHand[0]==11 else("Q" if self.playerHand[0]==12 else "K")))))
        for card in self.playerHand[1:]:
            m+=(","+str(card) if (card!=1 and card<11) else(",A" if card==1 else(",J" if card==11 else(",Q" if card==12 else ",K"))))
        m+="]"
        m+=("C["+str(self.playerChip)+"]")
        m+=("BET["+str(self.bet)+"]")
        return m

    def getMsg(self,fullMsg=False):
        m="=========================\n"
        m+="    Banker's Hand:"
        if not self.stand:
            m+=" ?"
            for card in self.bankerHand[1:]:
                m+=(" "+(str(card) if (card!=1 and card<11) else("A" if card==1 else("J" if card==11 else("Q" if card==12 else "K")))))
        else:
            for card in self.bankerHand:
                m+=(" "+(str(card) if (card!=1 and card<11) else("A" if card==1 else("J" if card==11 else("Q" if card==12 else "K")))))
        m+=("\n  > Banker's Sum: "+str(self.bankerHandSum))
        
        m+="\n    Player's Hand:"
        for card in self.playerHand:
            m+=(" "+(str(card) if (card!=1 and card<11) else("A" if card==1 else("J" if card==11 else("Q" if card==12 else "K")))))
        m+=("\n  > Player's Sum: "+str(self.playerHandSum))
        
        if fullMsg:
            m+=("\n    Player's Remainig chip: "+str(self.playerChip)+"\n    Bet: "+str(self.bet))
        
        if self.debug:
            m+=("\n    MSG: "+self.getStatus())
            m+=("\n    Deck Length: "+str(len(self.deck)))
            m+=("\n    Deck: "+str(self.deck))

        m+="\n=========================\n"
        return m

    def getStand(self):
        return self.stand

    def getEnded(self):
        return self.ended

    def getGameResult(self):
        return self.gameResult
        
    def calcBankerSum(self):
        if self.stand:
            cards=self.bankerHand
            sums=[0]
            for card in cards:
                toadd=card
                if card in {11,12,13}:
                    toadd=10
                if card!=1:
                    for i in range(len(sums)):
                        sums[i]+=toadd
                else:
                    sums=[i+j for i in sums for j in [1,11]]
            if len(sums)>1:
                sums.sort()
                if sums[1]>21:
                    sums=sums[:1]
                else:
                    sums=sums[:2]
            self.bankerHandSum=sums

    def calcPlayerSum(self):
        cards=self.playerHand
        sums=[0]
        for card in cards:
            toadd=card
            if card in {11,12,13}:
                toadd=10
            if card!=1:
                for i in range(len(sums)):
                    sums[i]+=toadd
            else:
                sums=[i+j for i in sums for j in [1,11]]
        if len(sums)>1:
            sums.sort()
            if sums[1]>21:
                sums=sums[:1]
            else:
                sums=sums[:2]
        self.playerHandSum=sums

    def hit(self):
        return self.deck.pop(0)

    def bankerHit(self):
        print("\nBanker Hit")
        self.bankerHand.append(self.hit())
        self.calcBankerSum()
        print(self.getMsg())
        if self.bankerHandSum!="?" and max(self.bankerHandSum)>=21:
            self.judge()

    def playerHit(self):
        print("\nPlayer Hit")
        self.playerHand.append(self.hit())
        self.calcPlayerSum()
        print(self.getMsg())
        if max(self.playerHandSum)>=21:
            self.judge()

    def bankerHitsTo17(self):
        while(max(self.bankerHandSum)<17):
            self.bankerHit()

    def playerStand(self):
        print("\nPlayer Stand")
        self.stand=True
        self.calcBankerSum()
        print(self.getMsg())
        self.bankerHitsTo17()
        if not self.ended:
            self.judge()

    def playerDouble(self):
        print("\nPlayer Double")
        self.bet*=2
        self.playerHit()
        if(not self.ended):
            self.playerStand()

    def judge(self):
        print("\n>>> Game Result <<<")
        self.stand=True
        self.calcBankerSum()
        self.calcPlayerSum()
        b=max(self.bankerHandSum)
        p=max(self.playerHandSum)
        print(self.getMsg(fullMsg=True))
        if b>21:
            print("Banker bust")
            self.playerWin()
        else:
            if p>21:
                print("Player bust")
                self.playerLose()
            else:
                self.playerWin() if p>b else(self.playerLose() if p<b else self.gamePush())
        self.ended=True

    def playerWin(self):
        print("Player Win!")
        self.playerChip+=self.bet
        self.gameResult=1
        print("Remaining chip: ",self.playerChip,"\n")
    
    def playerLose(self):
        print("Player Lose!")
        self.playerChip-=self.bet
        self.gameResult=-1
        print("Remaining chip: ",self.playerChip,"\n")

    def gamePush(self):
        print("PUSH!")
        self.gameResult=0
        print("Remaining chip: ",self.playerChip,"\n")